### PR TITLE
Update the transaction factories & unit tests fixes

### PR DESCRIPTION
**Pull request by popenta**
Updated all the classes that are using the `Transaction` class and fixed the unit tests.

### CODE REVIEW

File: multiversx_sdk_core/transaction_builders/relayed_v2_builder_test.py [hunk 1]
```diff
@@ -32,18 +32,19 @@ def test_without_arguments(self):
...
         inner_tx = Transaction(
             chain_id=network_config.chain_id,
-            sender=Address.from_bech32(self.alice.label),
-            receiver=Address.from_bech32("erd1qqqqqqqqqqqqqqqpqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzllls8a5w6u"),
+            sender=self.alice.label,
+            receiver="erd1qqqqqqqqqqqqqqqpqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzllls8a5w6u",
             gas_limit=10,
             nonce=15,
-            data=TransactionPayload.from_str("getContractConfig")
+            data=TransactionPayload.from_str("getContractConfig").data
```
Comment (andreibancioiu) [inline multiversx_sdk_core/transaction_builders/relayed_v2_builder_test.py:43]: Yup. I think `TransactionPayload` should be removed. It's not captured within the specs, and indeed, it should not be. We'll still keep it for some time in js-core, but can we perhaps remove it from py-core?

File: multiversx_sdk_core/transaction_factories/transaction_builder.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from multiversx_sdk_core.transaction import Transaction
+
+
+class IConfig(Protocol):
+    chain_id: str
+    min_gas_limit: int
+    gas_limit_per_byte: int
+
+
+class TransactionBuilder:
```
Comment (andreibancioiu) [inline multiversx_sdk_core/transaction_factories/transaction_builder.py:14]: Until `transaction_builders` gets removed (v2), we'll have 2 `TransactionBuilder` objects. Hopefully, it will be all right.

File: multiversx_sdk_core/custom_tokens.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+from multiversx_sdk_core.codec import decode_unsigned_number
+from multiversx_sdk_core.errors import InvalidTokenIdentifierError
+
```
Comment (andreibancioiu) [inline multiversx_sdk_core/custom_tokens.py:3]: Now I see. Perhaps we should rename the file to `tokens`, in `sdk-specs`, as well?

File: multiversx_sdk_core/custom_tokens.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+class TokenTransfer:
+    def __init__(self, token: Token, amount: int) -> None:
+        """`amount` should always be in atomic units: 1 EGLD = 1000000000000000000"""
+        self.token = token
+        self.amount = amount
+
+
+# the rest of the methods will be implemented in a future PR
```
Comment (andreibancioiu) [inline multiversx_sdk_core/custom_tokens.py:18]: In the future PR, we should add unit tests, as well.

File: multiversx_sdk_core/custom_tokens.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return token.nonce == 0
+
+    def extract_nonce_from_extended_identifier(self, identifier: str) -> int:
+        parts = identifier.split("-")
+
+        if len(parts) != 3:
+            raise InvalidTokenIdentifierError("You have not provided the extended identifier")
+
+        # check length of random sequence
+        if len(parts[1]) != 6:
```
Comment (andreibancioiu) [inline multiversx_sdk_core/custom_tokens.py:33]: Perhaps we can also define some private methods in order to reuse some validation logic?

File: multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        return transaction
+
+    def create_transaction_for_execute(self,
+                                       sender: IAddress,
+                                       contract: IAddress,
+                                       function: str,
+                                       gas_limit: int,
+                                       arguments: List[Any] = [],
+                                       native_transfer_amount: int = 0,
+                                       token_transfers: List[TokenTransfer] = []) -> Transaction:
```
Comment (andreibancioiu) [inline multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py:69]: Oh, nice, so you've also implemented `transfer & execute` :rocket:

Let's though extract this new code to a separate PR, to be separately reviewed.

File: multiversx_sdk_core/transaction_builders/esdt_builders.py [hunk 7]
```diff
@@ -70,6 +71,6 @@ def _build_payload_parts(self) -> List[str]:
             *(args_to_strings(["canMint", "true"]) if self.can_mint else []),
             *(args_to_strings(["canBurn", "true"]) if self.can_burn else []),
             *(args_to_strings(["canChangeOwner", "true"]) if self.can_change_owner else []),
-            *(args_to_strings(["canUpgrade", "true"]) if self.can_upgrade else []),
-            *(args_to_strings(["canAddSpecialRoles", "true"]) if self.can_add_special_roles else [])
+            *(args_to_strings(["canUpgrade", str(self.can_upgrade).lower()])),
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk_core/transaction_builders/esdt_builders.py:74]: I've sent a transaction explicitly setting `canUpgrade` to `False` and it works. I don't think that the constructor needs to be adjusted because we don't want any change since this will be deprecated in the next release.

File: multiversx_sdk_core/transaction_builders/relayed_v2_builder_test.py (hunk 1 above)
Comment (popenta) [reply multiversx_sdk_core/transaction_builders/relayed_v2_builder_test.py:43]: It's now removed from the tests.

File: multiversx_sdk_core/transaction_factories/token_management_transactions_factory.py [hunk 8]
```diff
@@ -32,12 +33,19 @@ class IConfig(Protocol):
     esdt_contract_address: IAddress
 
 
-class TokenManagementTransactionIntentsFactory:
+class RegisterAndSetAllRolesTokenType(Enum):
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk_core/transaction_factories/token_management_transactions_factory.py:36]: Indeed. It's now exported.

File: multiversx_sdk_core/custom_tokens.py (hunk 3 above)
Comment (popenta) [reply multiversx_sdk_core/custom_tokens.py:3]: Renamed to `tokens.py`. Will follow up with a PR on `sdk-specs`.

File: multiversx_sdk_core/custom_tokens.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+class Token:
+    def __init__(self, identifier: str = "", nonce: int = 0) -> None:
+        self.identifier = identifier
+        self.nonce = nonce
+
+
+class TokenTransfer:
+    def __init__(self, token: Token, amount: int) -> None:
+        """`amount` should always be in atomic units: 1 EGLD = 1000000000000000000"""
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk_core/custom_tokens.py:13]: Now using the example from the specs.

File: multiversx_sdk_core/custom_tokens.py (hunk 4 above)
Comment (popenta) [reply multiversx_sdk_core/custom_tokens.py:18]: Indeed, will add tests.

File: multiversx_sdk_core/custom_tokens.py (hunk 5 above)
Comment (popenta) [reply multiversx_sdk_core/custom_tokens.py:33]: Indeed. Did both suggestions.

File: multiversx_sdk_core/transaction_factories/smart_contract_transaction_factory_test.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+from multiversx_sdk_core.address import Address
+from multiversx_sdk_core.constants import CONTRACT_DEPLOY_ADDRESS
+from multiversx_sdk_core.custom_tokens import Token, TokenTransfer
+from multiversx_sdk_core.transaction_factories.smart_contract_transactions_factory import \
+    SmartContractTransactionsFactory
+from multiversx_sdk_core.transaction_factories.transactions_factory_config import \
+    TransactionsFactoryConfig
+
+
+class TestSmartContract:
```
[the remark this answers is not in the document]
Comment (popenta) [reply multiversx_sdk_core/transaction_factories/smart_contract_transaction_factory_test.py:12]: Renamed.

File: multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py (hunk 6 above)
Comment (popenta) [reply multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py:69]: Extracted this code, adjusted old code, will follow-up with a PR containing this code.

**Review by andreibancioiu — APPROVED** [review]

File: multiversx_sdk_core/transaction_factories/transaction_builder.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self,
+                 config: IConfig,
+                 sender: IAddress,
+                 receiver: IAddress,
+                 data_parts: List[str],
+                 gas_limit: int,
+                 add_data_movement_gas: bool,
```
Comment (ssd04) [inline multiversx_sdk_core/transaction_factories/transaction_builder.py:21]: [optional] maybe rename this param? at least for me it's not very clear from the name what it enables.
But i also cannot think of a better name

**Review by ssd04 — APPROVED** [review]

File: multiversx_sdk_core/transaction_factories/transaction_builder.py (hunk 11 above)
Comment (popenta) [reply multiversx_sdk_core/transaction_factories/transaction_builder.py:21]: I've named it the same as in `js-core`, but if we think of a better name we will change it in a future PR.
